src/Featurizers/LagLeadOperatorFeaturizer.h

```diff
@@ -0,0 +1,224 @@
+///                 Input type is a tuple of std::vector<std::string>, representing grains, and InputT, representing target column
+///
+template <typename InputT>
+class LagLeadOperatorTransformer :
+    public Transformer<
+        InputT,
+        Microsoft::Featurizer::RowMajMatrix<nonstd::optional<InputT>>
+    > {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+
+    using BaseType =
+        Transformer<
+            InputT,
+            Microsoft::Featurizer::RowMajMatrix<nonstd::optional<InputT>>
+        >;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    LagLeadOperatorTransformer(std::uint32_t horizon, std::vector<std::int64_t> offsets);
+    LagLeadOperatorTransformer(Archive &ar);
+
+    ~LagLeadOperatorTransformer(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(LagLeadOperatorTransformer);
+
+    bool operator==(LagLeadOperatorTransformer const &other) const;
+    bool operator!=(LagLeadOperatorTransformer const &other) const;
+
+    void save(Archive &ar) const override;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Types
+    // |
+    // ----------------------------------------------------------------------
+    using GrainType  = std::vector<std::string>;
+    using TargetType = nonstd::optional<InputT>;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    std::uint32_t const                     _horizon;
+    std::vector<std::int64_t> const         _offsets;
+    std::int64_t const                      _max_future_offset;
+    std::int64_t const                      _max_prior_offset;
+    Components::CircularBuffer<TargetType>  _buffer;
+    
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+    // MSVC runs into problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        execute_helper(nonstd::optional<typename BaseType::InputType>(input), callback);
+    }
+
+    void flush_impl(typename BaseType::CallbackFunction const & callback) override {
+        std::int64_t _numPending = 0;
+        while (_numPending < _max_future_offset) {
+            execute_helper(Traits<nonstd::optional<InputT>>::CreateNullValue(), callback);
```

Comment:
>execute_helper [](start = 12, length = 14)

Can this call `execute` directly? That would eliminate the need of `execute_helper`.

Reply:
I don't think so because say the target has type std::string, execute_impl would have input type as std::string while flush_impl() need a method that takes in nonstd::optional<std::string> to pass a null value.